From a89bca2cc32887c3a8cfb8434aed5ca9a0976c17 Mon Sep 17 00:00:00 2001
From: Lingyun Zhu <lingyun.zhu@intel.com>
Date: Tue, 20 Nov 2012 19:24:35 +0800
Subject: Graphics:Add error check when create GraphicBuffer

BZ: 50141

Port JB 4.1 patch 60108 to JB 4.2.1

Category: aosp improvement
Domain:  Graphics.UI
Origin: internal
Upstream-Candiate:yes

Change-Id: Id89f3ba3708fdd2a8a8472aac5367e99c939465f
Orig-Change-Id: I608d9a9028dec9562c2b58a7b9b6a370404129da
Signed-off-by: Lingyun Zhu <lingyun.zhu@intel.com>
Reviewed-on: http://android.intel.com:8080/74864
Reviewed-by: Yao, Yong <yong.yao@intel.com>
Tested-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: Tong, BoX <box.tong@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Signed-off-by: binchen1 <bin.chen@intel.com>
---
 libs/gui/IGraphicBufferProducer.cpp |    6 +++++-
 libs/ui/GraphicBuffer.cpp           |    2 ++
 2 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/libs/gui/IGraphicBufferProducer.cpp b/libs/gui/IGraphicBufferProducer.cpp
index fc86e60..0f461e5 100644
--- a/libs/gui/IGraphicBufferProducer.cpp
+++ b/libs/gui/IGraphicBufferProducer.cpp
@@ -60,7 +60,11 @@ public:
         bool nonNull = reply.readInt32();
         if (nonNull) {
             *buf = new GraphicBuffer();
-            reply.read(**buf);
+            result = reply.read(**buf);
+            if(result != NO_ERROR) {
+                (*buf).clear();
+                return result;
+            }
         }
         result = reply.readInt32();
         return result;
diff --git a/libs/ui/GraphicBuffer.cpp b/libs/ui/GraphicBuffer.cpp
index 0ecd3d9..96a7188 100644
--- a/libs/ui/GraphicBuffer.cpp
+++ b/libs/ui/GraphicBuffer.cpp
@@ -282,6 +282,8 @@ status_t GraphicBuffer::unflatten(
     if (handle != 0) {
         status_t err = mBufferMapper.registerBuffer(handle);
         if (err != NO_ERROR) {
+            width = height = stride = format = usage = 0;
+            handle = NULL;
             ALOGE("unflatten: registerBuffer failed: %s (%d)",
                     strerror(-err), err);
             return err;
-- 
1.7.4.1

